modules/platforms/cpp/core/src/impl/ignite_environment.cpp

```diff
@@ -39,7 +39,47 @@ namespace ignite
 {
     namespace impl
     {
-        typedef common::concurrent::SharedPointer<impl::cluster::ClusterNodeImpl> SP_ClusterNodeImpl;
+        namespace
+        {
+            /*
+             * Stores IgniteEnvironment shared pointers.
+             */
+            class IgniteEnvironmentSPHolder
+            {
+                CriticalSection ptrsLock;
+                std::set<SharedPointer<IgniteEnvironment>*> ptrs;
+            public:
+                IgniteEnvironmentSPHolder() :
+                    ptrsLock()
+                {
+                    // No-op.
+                }
+
+                void AddSP(SharedPointer<IgniteEnvironment>* penv)
+                {
+                    CsLockGuard mtx(ptrsLock);
+
+                    ptrs.insert(penv);
+                }
+
+                void DeleteSP(SharedPointer<IgniteEnvironment>* penv)
+                {
+                    CsLockGuard mtx(ptrsLock);
+
+                    ptrs.erase(penv);
+
+                    delete penv;
+                }
+
+                ~IgniteEnvironmentSPHolder()
+                {
+                    std::set<SharedPointer<IgniteEnvironment>*>::iterator it;
+                    for (it = ptrs.begin(); it != ptrs.end(); ++it)
+                        delete* it;
+                }
+
+            } EnvSPHolder;
```

Comment:
I'd avoid global states. Let's discuss, if we can store it in other place.

Reply:
Discussed.
The fix on java side is required.